File: client/src/app/models/characters/pacman.ts
import { Direction } from '@models/interfaces';
import {Character, MOVE_INTERVAL} from './character';
import {GameMap} from '@models/map/game-map';

export class Pacman extends Character {
  moved: boolean = false;
  angle: number = 0;

  constructor(gameMap: GameMap, x: number, y: number, cellSize: number, simulationMode: boolean) {
    super(gameMap, x, y, cellSize, MOVE_INTERVAL, simulationMode);
  }

  get hasMoved(): boolean {
    return this.moved;
  }

  updateFromServerData(x: number, y: number, direction: Direction, currentTime: number): void {
    super.updateFrom(x, y, direction, currentTime);
    
    if (direction.x === 1) this.angle = 0;
    else if (direction.x === -1) this.angle = Math.PI;
    else if (direction.y === 1) this.angle = Math.PI/2;
    else if (direction.y === -1) this.angle = -Math.PI/2;
    
    this.isMoving = true;
    this.moved = true;
  } 

  update(currentTime: number): void {
    if (!this.updatePosition(currentTime)) return;

    const nextX = this.gridX + this.nextDirection.x;
    const nextY = this.gridY + this.nextDirection.y;

    if (this.isValidPosition(nextX, nextY)) {
      this.direction = {...this.nextDirection};
    } else if (this.handleTeleport()) {
      this.lastMoveTime = currentTime + 250;
      return;
    }

    const newX = this.gridX + this.direction.x;
    const newY = this.gridY + this.direction.y;

    if (this.isValidPosition(newX, newY)) {
      this.lastMoveTime = currentTime;

      this.gridX = newX;
      this.gridY = newY;

      if (!this.moved && (this.direction.x !== 0 || this.direction.y !== 0)) {
        this.moved = true;
      }
    }
  
    if (this.direction.x === 1) this.angle = 0;
    else if (this.direction.x === -1) this.angle = Math.PI;
    else if (this.direction.y === 1) this.angle = Math.PI/2;
    else if (this.direction.y === -1) this.angle = -Math.PI/2;
  }

  draw(ctx: CanvasRenderingContext2D, currentTime: number): void {
    ctx.save();
    ctx.translate(this.displayX + this.cellSize/2, this.displayY + this.cellSize/2);
    ctx.rotate(this.angle);

    const mouthOpen = this.isMoving ? (Math.sin(Date.now() * 0.025) + 1) / 4 : 0.15;

    ctx.beginPath();
    ctx.fillStyle = '#FFFF00';
    ctx.arc(0, 0, this.cellSize/2 - 2, mouthOpen, 2 * Math.PI - mouthOpen);
    ctx.lineTo(0, 0);
    ctx.fill();

    ctx.restore();
  }

  private handleTeleport(): boolean {
    const [pointA, pointB] = this.gameMap.teleportPoints;
    const target = [pointA, pointB].find(p => this.gridX === p.x && this.gridY === p.y);
    
    if (target) {
      ({x: this.gridX, y: this.gridY} = target === pointA ? pointB : pointA);
      return true;
    }
    
    return false;
  }
}

File: server/models/characters/pacman.py
from models.map.game_map import GameMap
from models.interfaces import Direction

import numpy as np

class Pacman:
    def __init__(self, game_map: GameMap, x: int, y: int):
        self.is_game_over = False
        self.game_map = game_map
        self.last_move_time = 0
        
        self.gridX = x
        self.gridY = y
        
        self.next_direction = Direction(0, 0)
        self.direction = Direction(0, 0)
    
    def update(self, current_time: int, move_interval: int = 225) -> bool:
        if self.is_game_over:
            return False
        
        if current_time - self.last_move_time < move_interval:
            return False

        if self.game_map.is_walkable(self.gridX + self.next_direction.x, self.gridY + self.next_direction.y):
            self.direction = Direction(self.next_direction.x, self.next_direction.y)
        else:
            teleport_points = self.game_map.get_teleport_points()
            if not teleport_points or len(teleport_points) < 2:
                return False
        
            point_a, point_b = teleport_points[0], teleport_points[1]
            if (self.gridX, self.gridY) == (point_a[0], point_a[1]):
                self.gridX, self.gridY = point_b[0], point_b[1]
                self.last_move_time = current_time + 250    
                return True
            elif (self.gridX, self.gridY) == (point_b[0], point_b[1]):
                self.gridX, self.gridY = point_a[0], point_a[1]
                self.last_move_time = current_time + 250
                return True
            
        nextX = self.gridX + self.direction.x
        nextY = self.gridY + self.direction.y
        if self.game_map.is_walkable(nextX, nextY):
            self.last_move_time = current_time
    
            self.gridX = nextX
            self.gridY = nextY
        
            return True
        
        return False
    
    def set_direction(self, direction: Direction):
        self.next_direction = direction
    
    def set_game_over(self):
        self.next_direction = Direction(0, 0)
        self.is_game_over = True